compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -2347,6 +2350,61 @@ public void visit(BLangTupleDestructure tupleDeStmt, AnalyzerData data) {
         if (type.tag != TypeTags.SEMANTIC_ERROR) {
             checkTupleVarRefEquivalency(tupleDeStmt.pos, tupleDeStmt.varRef,
                                         tupleDeStmt.expr.getBType(), tupleDeStmt.expr.pos, data);
+            if (tupleDeStmt.expr.getKind() == LIST_CONSTRUCTOR_EXPR) {
+                analyzeListConstrExprInTupleDestructure(tupleDeStmt.varRef,
+                        (BLangListConstructorExpr) tupleDeStmt.expr);
+            }
+        }
+    }
+
+    private void analyzeListConstrExprInTupleDestructure(BLangTupleVarRef tupleVarRef,
+                                                         BLangListConstructorExpr listExpr) {
+        List<BLangExpression> exprs = listExpr.exprs;
+        List<BLangExpression> tupleVars = tupleVarRef.expressions;
+
+        int nonRestTupleVarIndex = 0;
+        int noOfNonRestVars = tupleVars.size();
+
+        for (BLangExpression expr: exprs) {
+            int remainingNonRestVars = noOfNonRestVars - nonRestTupleVarIndex;
+
+            if (remainingNonRestVars == 0) {
+                return;
+            }
+
+            // remainingNonRestVars > 0
+            if (expr.getKind() == NodeKind.LIST_CONSTRUCTOR_SPREAD_OP) {
+                BLangExpression spreadOpExpr = ((BLangListConstructorSpreadOpExpr) expr).expr;
+                int noOfSpreadMembers = switch (spreadOpExpr.getBType().tag) {
+                    case TypeTags.ARRAY -> {
+                        BArrayType spreadArrayType = (BArrayType) spreadOpExpr.getBType();
+                        yield spreadArrayType.size;
+                    }
+                    case TypeTags.TUPLE -> {
+                        BTupleType spreadTupleType = (BTupleType) spreadOpExpr.getBType();
+                        yield spreadTupleType.getTupleTypes().size();
+                    }
+                    default -> 0;
+                };
+
+                if (remainingNonRestVars < noOfSpreadMembers) {
+                    return;
+                }
+
+                nonRestTupleVarIndex = nonRestTupleVarIndex + noOfSpreadMembers;
+                continue;
+            }
+
+            BLangExpression tupleMemberVar = tupleVars.get(nonRestTupleVarIndex++);
+            if (expr.getKind() == LIST_CONSTRUCTOR_EXPR && tupleMemberVar.getKind() == TUPLE_VARIABLE_REF) {
+                analyzeListConstrExprInTupleDestructure(
+                        (BLangTupleVarRef) tupleMemberVar, (BLangListConstructorExpr) expr);
+                continue;
```

Comment:
May have to handle errors also accordingly.

Reply:
https://github.com/ballerina-platform/ballerina-lang/pull/40283#discussion_r1481183514